packages/nextjs/src/server/wrapServerComponentWithSentry.ts

```diff
@@ -0,0 +1,67 @@
+import { captureException, getCurrentHub, startTransaction } from '@sentry/core';
+import * as domain from 'domain';
+
+import type { ServerComponentContext } from '../common/types';
+
+/**
+ * Wraps an `app` directory server component with Sentry error instrumentation.
+ */
+export function wrapServerComponentWithSentry<F extends (...args: any[]) => any>(
+  appDirComponent: F,
+  context: ServerComponentContext,
+): F {
+  const { componentRoute, componentType } = context;
+
+  // Even though users may define server components as async functions, for the client bundles
+  // Next.js will turn them into synchronous functions and it will transform any `await`s into instances of the `use`
+  // hook. 🤯
+  return new Proxy(appDirComponent, {
+    apply: (originalFunction, thisArg, args) => {
+      return domain.create().bind(() => {
+        let maybePromiseResult;
+
+        const transaction = startTransaction({
+          op: 'function.nextjs',
+          name: `${componentType} Server Component (${componentRoute})`,
+          status: 'ok',
+          metadata: {
+            source: 'component',
+          },
+        });
+
+        const currentScope = getCurrentHub().getScope();
+        if (currentScope) {
+          currentScope.setSpan(transaction);
```

Comment:
I wonder if you throw some React.withProfiler calls around child components if they'll show up attached as spans!

Reply:
We could try it out at some point. Right now, `Sentry.withProfiler` is still littered with browser code which is a big nono in server components.